src/thread.rs

```diff
@@ -15,18 +17,24 @@ use rustc_middle::{
     ty::{self, Instance},
 };
 
+use crate::sync::SynchronizationState;
 use crate::*;
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub enum SchedulingAction {
     /// Execute step on the active thread.
     ExecuteStep,
+    /// Execute a scheduler's callback.
+    ExecuteCallback,
```

Comment:
So *all* callbacks are for timeout stuff? I feel like that's worth making more clear, the mechanism looks way more general than that. Maybe call it a `Timeout` or `Timer` instead of the rather undescriptive `Callback`?